File: examples/go/single_url.go
package main

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/joho/godotenv"
)

const (
	defaultActorID = "apple_yang/instagram-transcripts-scraper"
	apiBaseURL     = "https://api.apify.com/v2"
)

type apiResponse struct {
	Data runInfo `json:"data"`
}

type runInfo struct {
	ID               string `json:"id"`
	Status           string `json:"status"`
	DefaultDatasetID string `json:"defaultDatasetId"`
}

func main() {
	if err := runSingle(); err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
}

func runSingle() error {
	repoRoot, err := findRepoRoot()
	if err != nil {
		return err
	}
	_ = godotenv.Load(filepath.Join(repoRoot, ".env"))

	token, err := requireEnv("APIFY_TOKEN")
	if err != nil {
		return err
	}
	actorID := envOrDefault("APIFY_ACTOR_ID", defaultActorID)
	sessionID := os.Getenv("INSTAGRAM_SESSIONID")
	_ = envOrDefault("MAX_CONCURRENCY", "3")
	reelURL := os.Getenv("INSTAGRAM_REEL_URL")
	if len(os.Args) > 1 {
		reelURL = os.Args[1]
	}
	if strings.TrimSpace(reelURL) == "" {
		return errors.New("Error: INSTAGRAM_REEL_URL is required. Set it in .env or pass it as a CLI argument.")
	}

	fmt.Println("Actor ID:", actorID)
	fmt.Println("Running request...")

	client := &http.Client{Timeout: 10 * time.Minute}
	run, err := startActorRun(client, token, actorID, reelURL, sessionID)
	if err != nil {
		return err
	}
	if run.ID == "" {
		return errors.New("Error: Actor run response did not include an id.")
	}

	finishedRun, err := waitForRun(client, token, run.ID)
	if err != nil {
		return err
	}
	if finishedRun.Status != "SUCCEEDED" {
		return fmt.Errorf("Error: Actor run finished with status %s.", finishedRun.Status)
	}
	if finishedRun.DefaultDatasetID == "" {
		return errors.New("Error: Actor run completed without a defaultDatasetId.")
	}

	items, rawItems, err := fetchDatasetItems(client, token, finishedRun.DefaultDatasetID)
	if err != nil {
		return err
	}

	outputDir := filepath.Join(repoRoot, "sample-output", "generated")
	outputFile := filepath.Join(outputDir, "go-single-url-response.json")
	if err := os.MkdirAll(outputDir, 0o755); err != nil {
		return err
	}
	if err := os.WriteFile(outputFile, append(rawItems, '\n'), 0o644); err != nil {
		return err
	}

	text := firstString(items, "text")
	segments := firstSlice(items, "segments")
	errMsg := firstString(items, "errMsg")

	fmt.Println("Output path:", outputFile)
	fmt.Println("Items:", len(items))
	fmt.Println("Transcript characters:", len(text))
	fmt.Println("Segments:", len(segments))
	if errMsg != "" {
		fmt.Println("errMsg:", errMsg)
	}

	return nil
}

func startActorRun(client *http.Client, token, actorID, reelURL, sessionID string) (runInfo, error) {
	body, err := json.Marshal(map[string]string{
		"videoUrl":  reelURL,
		"sessionid": sessionID,
	})
	if err != nil {
		return runInfo{}, err
	}
	actorIDForURL := strings.ReplaceAll(actorID, "/", "~")
	url := fmt.Sprintf("%s/acts/%s/runs", apiBaseURL, actorIDForURL)

	var response apiResponse
	if err := doJSON(client, token, http.MethodPost, url, body, &response); err != nil {
		return runInfo{}, err
	}
	return response.Data, nil
}

func waitForRun(client *http.Client, token, runID string) (runInfo, error) {
	for attempt := 0; attempt < 150; attempt++ {
		url := fmt.Sprintf("%s/actor-runs/%s", apiBaseURL, runID)
		var response apiResponse
		if err := doJSON(client, token, http.MethodGet, url, nil, &response); err != nil {
			return runInfo{}, err
		}
		switch response.Data.Status {
		case "SUCCEEDED", "FAILED", "TIMED-OUT", "ABORTED":
			return response.Data, nil
		}
		time.Sleep(2 * time.Second)
	}
	return runInfo{}, errors.New("Error: Timed out waiting for Actor run to finish.")
}

func fetchDatasetItems(client *http.Client, token, datasetID string) ([]map[string]any, []byte, error) {
	url := fmt.Sprintf("%s/datasets/%s/items", apiBaseURL, datasetID)
	raw, err := doRaw(client, token, http.MethodGet, url, nil)
	if err != nil {
		return nil, nil, err
	}
	var items []map[string]any
	if len(bytes.TrimSpace(raw)) > 0 {
		if err := json.Unmarshal(raw, &items); err != nil {
			return nil, nil, err
		}
	}
	if items == nil {
		items = []map[string]any{}
	}
	pretty, err := json.MarshalIndent(items, "", "  ")
	if err != nil {
		return nil, nil, err
	}
	return items, pretty, nil
}

func doJSON(client *http.Client, token, method, url string, body []byte, target any) error {
	raw, err := doRaw(client, token, method, url, body)
	if err != nil {
		return err
	}
	if len(bytes.TrimSpace(raw)) == 0 {
		return nil
	}
	return json.Unmarshal(raw, target)
}

func doRaw(client *http.Client, token, method, url string, body []byte) ([]byte, error) {
	var reader io.Reader
	if body != nil {
		reader = bytes.NewReader(body)
	}
	request, err := http.NewRequest(method, url, reader)
	if err != nil {
		return nil, err
	}
	request.Header.Set("Authorization", "Bearer "+token)
	request.Header.Set("Content-Type", "application/json")

	response, err := client.Do(request)
	if err != nil {
		return nil, err
	}
	defer response.Body.Close()

	raw, err := io.ReadAll(response.Body)
	if err != nil {
		return nil, err
	}
	if response.StatusCode < 200 || response.StatusCode >= 300 {
		return nil, fmt.Errorf("HTTP %d from Apify API: %s", response.StatusCode, string(raw))
	}
	return raw, nil
}

func firstString(items []map[string]any, field string) string {
	for _, item := range items {
		if value, ok := item[field].(string); ok && value != "" {
			return value
		}
	}
	return ""
}

func firstSlice(items []map[string]any, field string) []any {
	for _, item := range items {
		if value, ok := item[field].([]any); ok {
			return value
		}
	}
	return []any{}
}

func requireEnv(name string) (string, error) {
	value := strings.TrimSpace(os.Getenv(name))
	if value == "" {
		return "", fmt.Errorf("Error: %s is required. Set it in .env or your environment.", name)
	}
	return value, nil
}

func envOrDefault(name, defaultValue string) string {
	value := strings.TrimSpace(os.Getenv(name))
	if value == "" {
		return defaultValue
	}
	return value
}

func findRepoRoot() (string, error) {
	current, err := os.Getwd()
	if err != nil {
		return "", err
	}
	for {
		if fileExists(filepath.Join(current, ".env.example")) && fileExists(filepath.Join(current, "sample-data")) {
			return current, nil
		}
		parent := filepath.Dir(current)
		if parent == current {
			break
		}
		current = parent
	}
	return "", errors.New("Error: Could not locate repository root.")
}

func fileExists(path string) bool {
	_, err := os.Stat(path)
	return err == nil
}

File: examples/rust/src/bin/single_url.rs
use anyhow::{anyhow, bail, Context, Result};
use reqwest::Client;
use serde::Deserialize;
use serde_json::{json, Value};
use std::env;
use std::fs;
use std::path::PathBuf;
use tokio::time::{sleep, Duration};

const DEFAULT_ACTOR_ID: &str = "apple_yang/instagram-transcripts-scraper";
const API_BASE_URL: &str = "https://api.apify.com/v2";

#[derive(Debug, Deserialize)]
struct ApiResponse<T> {
    data: T,
}

#[derive(Debug, Deserialize)]
struct RunInfo {
    id: Option<String>,
    status: Option<String>,
    #[serde(rename = "defaultDatasetId")]
    default_dataset_id: Option<String>,
}

#[tokio::main]
async fn main() -> Result<()> {
    let repo_root = find_repo_root()?;
    let _ = dotenvy::from_path(repo_root.join(".env"));

    let token = require_env("APIFY_TOKEN")?;
    let actor_id = env::var("APIFY_ACTOR_ID").unwrap_or_else(|_| DEFAULT_ACTOR_ID.to_string());
    let sessionid = env::var("INSTAGRAM_SESSIONID").unwrap_or_default();
    let _max_concurrency = env::var("MAX_CONCURRENCY").unwrap_or_else(|_| "3".to_string());
    let reel_url = env::args()
        .nth(1)
        .or_else(|| env::var("INSTAGRAM_REEL_URL").ok())
        .filter(|value| !value.trim().is_empty())
        .ok_or_else(|| anyhow!("Error: Pass an Instagram Reel URL as a CLI argument."))?;

    println!("Actor ID: {actor_id}");
    println!("Running request...");

    let client = Client::builder()
        .timeout(Duration::from_secs(600))
        .build()
        .context("Failed to create HTTP client")?;

    let run = start_actor_run(&client, &token, &actor_id, &reel_url, &sessionid).await?;
    let run_id = run
        .id
        .ok_or_else(|| anyhow!("Error: Actor run response did not include an id."))?;
    let finished_run = wait_for_run(&client, &token, &run_id).await?;
    let status = finished_run.status.unwrap_or_default();
    if status != "SUCCEEDED" {
        bail!("Error: Actor run finished with status {status}.");
    }
    let dataset_id = finished_run
        .default_dataset_id
        .ok_or_else(|| anyhow!("Error: Actor run completed without a defaultDatasetId."))?;

    let items = fetch_dataset_items(&client, &token, &dataset_id).await?;

    let output_dir = repo_root.join("sample-output/generated");
    let output_file = output_dir.join("rust-single-url-response.json");
    fs::create_dir_all(&output_dir).context("Failed to create output directory")?;
    fs::write(
        &output_file,
        format!("{}\n", serde_json::to_string_pretty(&items)?),
    )
    .context("Failed to write output file")?;

    let text = first_string(&items, "text").unwrap_or_default();
    let segments_count = first_array_len(&items, "segments");
    let err_msg = first_string(&items, "errMsg").unwrap_or_default();

    println!("Output path: {}", output_file.display());
    println!("Items: {}", items.len());
    println!("Transcript characters: {}", text.chars().count());
    println!("Segments: {segments_count}");
    if !err_msg.is_empty() {
        println!("errMsg: {err_msg}");
    }

    Ok(())
}

async fn start_actor_run(
    client: &Client,
    token: &str,
    actor_id: &str,
    reel_url: &str,
    sessionid: &str,
) -> Result<RunInfo> {
    let actor_id_for_url = actor_id.replace('/', "~");
    let url = format!("{API_BASE_URL}/acts/{actor_id_for_url}/runs");
    let body = json!({
        "videoUrl": reel_url,
        "sessionid": sessionid,
    });

    let response: ApiResponse<RunInfo> =
        send_json(client, token, reqwest::Method::POST, &url, Some(body)).await?;
    Ok(response.data)
}

async fn wait_for_run(client: &Client, token: &str, run_id: &str) -> Result<RunInfo> {
    for _ in 0..150 {
        let url = format!("{API_BASE_URL}/actor-runs/{run_id}");
        let response: ApiResponse<RunInfo> =
            send_json(client, token, reqwest::Method::GET, &url, None).await?;
        let status = response.data.status.as_deref().unwrap_or("");
        if matches!(status, "SUCCEEDED" | "FAILED" | "TIMED-OUT" | "ABORTED") {
            return Ok(response.data);
        }
        sleep(Duration::from_secs(2)).await;
    }
    bail!("Error: Timed out waiting for Actor run to finish.")
}

async fn fetch_dataset_items(client: &Client, token: &str, dataset_id: &str) -> Result<Vec<Value>> {
    let url = format!("{API_BASE_URL}/datasets/{dataset_id}/items");
    send_json(client, token, reqwest::Method::GET, &url, None).await
}

async fn send_json<T: for<'de> Deserialize<'de>>(
    client: &Client,
    token: &str,
    method: reqwest::Method,
    url: &str,
    body: Option<Value>,
) -> Result<T> {
    let mut request = client
        .request(method, url)
        .bearer_auth(token)
        .header("Content-Type", "application/json");
    if let Some(body) = body {
        request = request.json(&body);
    }

    let response = request.send().await.context("Apify API request failed")?;
    let status = response.status();
    let text = response.text().await.context("Failed to read Apify API response")?;
    if !status.is_success() {
        bail!("HTTP {} from Apify API: {}", status.as_u16(), text);
    }
    serde_json::from_str(&text).context("Failed to parse Apify API response JSON")
}

fn first_string(items: &[Value], field: &str) -> Option<String> {
    items
        .iter()
        .filter_map(|item| item.get(field)?.as_str())
        .find(|value| !value.is_empty())
        .map(ToString::to_string)
}

fn first_array_len(items: &[Value], field: &str) -> usize {
    items
        .iter()
        .filter_map(|item| item.get(field)?.as_array())
        .next()
        .map_or(0, Vec::len)
}

fn require_env(name: &str) -> Result<String> {
    let value = env::var(name).unwrap_or_default();
    if value.trim().is_empty() {
        bail!("Error: {name} is required. Set it in .env or your environment.");
    }
    Ok(value)
}

fn find_repo_root() -> Result<PathBuf> {
    let mut current = env::current_dir().context("Failed to read current directory")?;
    loop {
        if current.join(".env.example").exists() && current.join("sample-data").exists() {
            return Ok(current);
        }
        if !current.pop() {
            break;
        }
    }
    bail!("Error: Could not locate repository root.")
}
